services/motion/builtin/builtin.go

```diff
@@ -231,24 +236,109 @@ func (ms *builtIn) MoveOnGlobe(
+	var planMu sync.Mutex
+	var currentPlan [][]referenceframe.Input = nil
+
+	// Shared variable between threads that can trigger replanning and planning thread
+	var replanFlag atomic.Bool
+	replanFlag.Store(true)
+
+	var successFlag atomic.Bool
+	successFlag.Store(false)
+	errChan := make(chan error)
+
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	ms.cancelFn = cancelFn
+
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+
+		positionPollingTicker := time.NewTicker(time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond)
+		obstaclePollingTicker := time.NewTicker(time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond)
+
+		for !successFlag.Load() && ctx.Err() == nil {
+			select {
+			case <-positionPollingTicker.C:
+				// TODO: theres nothing stopping us from making another thread before the previous has completed
+				ms.backgroundWorkers.Add(1)
+				goutils.ManagedGo(func() {
+					fmt.Println("position poll")
+					current, _, err := movementSensor.Position(cancelCtx, nil)
+					if err != nil {
+						errChan <- err
+						return
+					}
+					if spatialmath.GeoPointToPose(current, destination).Point().Norm() <= kinematicsOptions.PlanDeviationThresholdMM {
+						successFlag.Store(true)
+					}
+
+					// TODO: the function that actually monitors position - and pass context in
+				}, ms.backgroundWorkers.Done)
+			case <-obstaclePollingTicker.C:
+				// TODO: theres nothing stopping us from making another thread before the previous has completed
+				ms.backgroundWorkers.Add(1)
+				goutils.ManagedGo(func() {
+					fmt.Println("obstacle poll")
+					replanFlag.Store(true)
+
+					// this is gonna get removed when I pass context along the way
+					if !utils.SelectContextOrWait(cancelCtx, 1*time.Millisecond) {
+						fmt.Println("I've been cancelled")
+					}
+
+					// TODO: the function that actually monitors obstacles - and pass context in
+				}, ms.backgroundWorkers.Done)
+			default:
+			}
+			if replanFlag.Load() {
+				fmt.Println("replanning")
+
+				ms.cancelFn()
+				cancelCtx, cancelFn = context.WithCancel(context.Background())
+				ms.cancelFn = cancelFn
+
+				plan, kb, err := ms.planMoveOnGlobe(
+					ctx,
+					componentName,
+					destination,
+					movementSensor,
+					obstacles,
+					kinematicsOptions,
+					extra,
+				)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				planMu.Lock()
+				currentPlan = plan
+				kinematicBase = kb
+				planMu.Unlock()
+
+				replanFlag.Store(false)
+			}
+		}
+	}, ms.backgroundWorkers.Done)
 
 	// execute the plan
-	for i := 1; i < len(plan); i++ {
-		ms.logger.Info(plan[i])
-		if err := kb.GoToInputs(ctx, plan[i]); err != nil {
-			return false, err
+	for !successFlag.Load() {
+		for i := 1; len(currentPlan) > 1; {
+			ms.logger.Info(currentPlan[i])
+			if err := kinematicBase.GoToInputs(cancelCtx, currentPlan[i]); err != nil {
```

Comment:
i think we need to handle the error case more carefully here. my reasoning is that sometimes there will be a non-nil error because we have requested a replan. in this case, we should not return false and err

Reply:
Yeah thats a really good point.  I think the answer may lie with catching that error case explicitly but I'll have to think on it more